cpp/src/tsne/fft_tsne.cuh

```diff
@@ -0,0 +1,525 @@
+        (value_idx)NTHREADS_128);
+      FFT::compute_potential_indices<value_idx, value_t, n_terms,
+                                     n_interpolation_points>
+        <<<num_blocks, NTHREADS_128, 0, stream>>>(
+          potentialsQij_device.data(), point_box_idx_device.data(),
+          y_tilde_values.data(), x_interpolated_values_device.data(),
+          y_interpolated_values_device.data(), n, n_boxes_per_dim);
+    }
+
+    value_t normalization;
+    {
+      // Compute repulsive forces
+      // Make the negative term, or F_rep in the equation 3 of the paper.
+      auto num_blocks = raft::ceildiv(n, (value_idx)NTHREADS_1024);
+      FFT::compute_repulsive_forces_kernel<<<num_blocks, NTHREADS_1024, 0,
+                                             stream>>>(
+        repulsive_forces_device.data(), normalization_vec_device.data(), Y,
+        Y + n, potentialsQij_device.data(), n, n_terms);
+
+      auto norm_vec_thrust =
+        thrust::device_pointer_cast(normalization_vec_device.data());
+
+      value_t sumQ =
+        thrust::reduce(thrust_policy, norm_vec_thrust,
+                       norm_vec_thrust + normalization_vec_device.size(), 0.0f,
+                       thrust::plus<value_t>());
+      normalization = sumQ - n;
+    }
+
+    // Compute attractive forces
+    {
+      auto num_blocks = raft::ceildiv(NNZ, (value_idx)NTHREADS_1024);
+      FFT::compute_Pij_x_Qij_kernel<<<num_blocks, NTHREADS_1024, 0, stream>>>(
+        attractive_forces_device.data(), VAL, ROW, COL, Y, n, NNZ);
+    }
+
+    // Apply Forces
+    {
+      auto num_blocks = mp_count * integration_kernel_factor;
+
+      FFT::IntegrationKernel<<<num_blocks, NTHREADS_1024, 0, stream>>>(
+        Y, attractive_forces_device.data(), repulsive_forces_device.data(),
+        gains_device.data(), old_forces_device.data(), learning_rate,
+        normalization, momentum, exaggeration, n);
+    }
+
+    auto att_forces_thrust =
+      thrust::device_pointer_cast(attractive_forces_device.data());
+    auto old_forces_thrust =
+      thrust::device_pointer_cast(old_forces_device.data());
+
+    thrust::transform(thrust_policy, old_forces_thrust, old_forces_thrust + n,
+                      att_forces_thrust, FunctionalSquare());
+
+    thrust::transform(thrust_policy, att_forces_thrust, att_forces_thrust + n,
+                      att_forces_thrust + n, att_forces_thrust,
+                      thrust::plus<value_t>());
+
+    thrust::transform(thrust_policy, att_forces_thrust,
+                      att_forces_thrust + attractive_forces_device.size(),
+                      att_forces_thrust, FunctionalSqrt());
+
+    value_t grad_norm =
+      thrust::reduce(thrust_policy, att_forces_thrust,
+                     att_forces_thrust + attractive_forces_device.size(), 0.0f,
+                     thrust::plus<value_t>()) /
+      attractive_forces_device.size();
+
+    if (grad_norm <= min_grad_norm) {
+      std::cout << "BREAKING EARLY" << std::endl;
```

Comment:
Should probably use CUML_LOG_DEBUG for this one, and maybe a more friendly message